### DISCUSSION TITLE
Use centre of viewport for picking instead of cursor?

### DISCUSSION THREAD

**Discussion by 66OJ66** [aevyrie/bevy_mod_picking#165] [Q&A] [answered]
Hi,

In games where the cursor is hidden (e.g. FPS), it's common to look at objects in order to pick/interact with them.

Based on this issue (https://github.com/bevyengine/bevy/issues/2645) I don't think it's possible to lock the cursor to the centre of the screen currently, so is there a way to use the viewport centre instead of the cursor for picking objects?

For example, in the video below I'd want to pick the cube anytime I hover over it with the white square in the centre of the screen.

https://user-images.githubusercontent.com/59021291/194153300-9a3dee55-ca0c-4eab-9132-1f05f2ebfda8.mp4

**Comment by aevyrie** [comment]
Basically, instead of constantly moving the camera ray to match the mouse, all you have to do is set it to the center of the screen and leave it alone.

I'd recommend using bevy_mod_raycast so you can get the items under the reticle and the mouse seperately, but if you want something quick and dirty that works, I think the easiest way to do this is to simply overwrite the value that the plugin adds from the mouse:

```rs
// When you add this system in your app make sure it is set to run 
//    after `PickingSystem::UpdatePickSourcePositions` 
//    AND before `PickingSystem::BuildRays`
fn reset_pointer(
     mut pick_source_query: Query<&mut PickingCamera>,
) {
    pick_source.cast_method = RayCastMethod::Transform;
}
```

**Comment by 66OJ66** [comment, marked as answer]
Hi - thanks for your reply!

I tried the quick method you mentioned, but it didn't produce the desired result by itself.

I ended up doing 2 things:
1) Modify the `PickingCameraBundle` to use `RayCastMethod::Transform` as the cast_method before adding it to my Camera entity - e.g
``` rust
let mut picking_camera_bundle = PickingCameraBundle::default();
picking_camera_bundle.source.cast_method = RayCastMethod::Transform;
```

2) Make a custom version of `PickingPlugin` which excludes the `update_pick_source_positions` system.

Code below:
``` rust
use bevy::prelude::*;
use bevy::app::PluginGroupBuilder;
use bevy::ecs::schedule::ShouldRun;
use bevy_mod_picking::{DebugCursorPickingPlugin, HighlightablePickingPlugins, InteractablePickingPlugin, PickingPluginsState, PickingRaycastSet, PickingSystem};

/// A custom version of DefaultPickingPlugins.
pub(super) struct CustomPickingPlugins;

impl PluginGroup for CustomPickingPlugins {
    fn build(&mut self, group: &mut PluginGroupBuilder) {
        group.add(CustomPickingPlugin);
        group.add(InteractablePickingPlugin);
        group.add(DebugCursorPickingPlugin);
        HighlightablePickingPlugins.build(group);
    }
}

/// A custom version of PickingPlugin which doesn't change the raycast direction.
struct CustomPickingPlugin;

impl Plugin for CustomPickingPlugin {
    fn build(&self, app: &mut App) {
        app.init_resource::<PickingPluginsState>()
            .add_system_set_to_stage(
                CoreStage::First,
                SystemSet::new()
                    .with_run_criteria(|state: Res<PickingPluginsState>| {
                        simple_criteria(state.enable_picking)
                    })
                    // The update_pick_source_positions system is excluded here
                    .with_system(
                        bevy_mod_raycast::build_rays::<PickingRaycastSet>
                            .label(PickingSystem::BuildRays)
                            .before(PickingSystem::UpdateRaycast),
                    )
                    .with_system(
                        bevy_mod_raycast::update_raycast::<PickingRaycastSet>
                            .label(PickingSystem::UpdateRaycast)
                            .before(PickingSystem::UpdateIntersections),
                    )
                    .with_system(
                        bevy_mod_raycast::update_intersections::<PickingRaycastSet>
                            .label(PickingSystem::UpdateIntersections),
                    ),
            );
    }
}

fn simple_criteria(flag: bool) -> ShouldRun {
    if flag {
        ShouldRun::Yes
    } else {
        ShouldRun::No
    }
}
```

Just out of interest, I noticed that `cast_method` in `RayCastSource` defaults to `RayCastMethod::Screenspace(Vec2::ZERO)` (i.e. the bottom-left of the screen) - is there a rationale behind using that instead of `RayCastMethod::Transform` as the default?

**Reply by aevyrie** [reply]
> is there a rationale behind using that instead of RayCastMethod::Transform as the default?

No rational that I can recall.

I was going to suggest the exact plugin wrangling you ended up doing, but thought it might be confusing if you were a bevy beginner. Glad you figured it out. :slightly_smiling_face: 

Is there some reason you are using mod_picking for this? I would've expected bevy_mod_raycast would be a much better fit. mod_picking is pretty opinionated and wasn't built with this use case in mind.

**Reply by 66OJ66** [reply]
> Is there some reason you are using mod_picking for this? I would've expected bevy_mod_raycast would be a much better fit. mod_picking is pretty opinionated and wasn't built with this use case in mind.

I opted for mod_picking because it has `PickingEvent` and `DebugCursorPickingPlugin`.
Both of these provided functionality I needed, so it made more sense just to use mod_picking and make some slight adjustments instead.

I'll mark this as answered now - thank you for your help 🙂

**Reply by aevyrie** [reply]
FYI, the debug cursor is supplied by mod_raycast.
